### DISCUSSION TITLE
Provide a multiarch build (armhf/arm64) for the container

### DISCUSSION THREAD

**Discussion by georgr** [developmentseed/titiler#387] [Feature Request]
Please provide a multiarch build (armhf/arm64) for the container.

Awesome work!

**Comment by vincentsarago** [comment]
@georgr Please feel free to start a PR if this is something you want/need. 

see https://github.com/vincentsarago/containers/pkgs/container/uvicorn-gunicorn

**Comment by vincentsarago** [comment]
@georgr sadly rasterio do not yet comes with wheels for arm64 so you'll need to have gdal installed first. I'm going to wait for rasterio wheels to be provided for the arm64 arch. 

While this I'm going to add `platform: linux/amd64` in docker-compose.yml and make sure we tag the image with it

**Reply by georgr** [reply]
@vincentsarago 

I had success with following Dockerfile. The Dockerfile is by no means "ready for production", I think the size can be reduced and I do not get the idea why the root certificates has to be copied in the container.

Dockerfile:
```
FROM python:3.8 as builder

RUN mkdir /build
WORKDIR /build


# Ensure root certificates are always updated at evey container build
# and curl is using the latest version of them


COPY src/titiler/ /tmp/titiler/

RUN apt-get update
RUN apt-get install -y python3-numpy gdal-bin libgdal-dev

RUN mkdir /usr/local/share/ca-certificates/cacert.org
RUN cd /usr/local/share/ca-certificates/cacert.org && curl -k -O https://www.cacert.org/certs/root.crt 
RUN cd /usr/local/share/ca-certificates/cacert.org && curl -k -O https://www.cacert.org/certs/class3.crt
RUN update-ca-certificates
ENV CURL_CA_BUNDLE /etc/ssl/certs/ca-certificates.crt

RUN pip install --prefix=/build /tmp/titiler/core /tmp/titiler/mosaic /tmp/titiler/application --no-cache-dir
RUN pip install --prefix=/build uvicorn

RUN rm -rf /tmp/titiler

FROM python:3.8-slim as base

COPY --from=builder /build /usr/local

RUN apt-get update
RUN apt-get install -y python3-numpy gdal-bin libgdal-dev
RUN apt-get -y install curl

RUN mkdir /usr/local/share/ca-certificates/cacert.org
RUN cd /usr/local/share/ca-certificates/cacert.org && curl -k -O https://www.cacert.org/certs/root.crt 
RUN cd /usr/local/share/ca-certificates/cacert.org && curl -k -O https://www.cacert.org/certs/class3.crt
RUN update-ca-certificates
ENV CURL_CA_BUNDLE /etc/ssl/certs/ca-certificates.crt

ENV MODULE_NAME titiler.application.main
ENV VARIABLE_NAME app

CMD ["uvicorn", "titiler.application.main:app", "--host", "0.0.0.0", "--port", "80"]
```

I´m building the container with buildx:

```docker buildx build --push --platform linux/arm64/v8,linux/amd64 --tag georam/titiler . ``` 

If you have an arm64 platform available you can either build the container yourself with the provided Dockerfile or test the container pushed on [docker hub](https://hub.docker.com/r/georam/titiler).

A major issue is in my opinion the build time for arm64 with about 20 minutes (on my intel platform) and over an hour for arm64 and armhf combined. 

I´m not sure if this approach is feasible with this build times, but titiler is running on my k3s pi4 development cluster like a charm 😀 

Do you have any thoughts on this?

**Reply by vincentsarago** [reply]
In tests we never really had issue with the image size, about build time, it's really weird, it shouldn't take more than 2 minutes.

Note:
in 
```
RUN pip install --prefix=/build /tmp/titiler/core /tmp/titiler/mosaic /tmp/titiler/application --no-cache-dir
```

you should do 
```
RUN pip install --prefix=/build /tmp/titiler/core /tmp/titiler/mosaic /tmp/titiler/application --no-cache-dir --no-binary rasterio
```

**Reply by f-skold** [reply]
I have seen similar difference in build times when porting a custom docker images with PHP.
If you build the same image on an Arm64 host (on for example AWS), build time was in the same range as building the image natively on the AMD64/x86_64.

@vincentsarago, I think I had to comment out the platform directive in the docker-compose.yml, to get the file to work on my machine:

```
services:
  titiler:
    # See https://github.com/developmentseed/titiler/discussions/387
    # platform: linux/amd64
 ````

Version information:
```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.4 LTS
Release:        20.04
Codename:       focal
$ docker --version
Docker version 20.10.16, build aa7e414
$ docker-compose --version
docker-compose version 1.26.0, build d4451659
```

**Comment by Firefishy** [comment]
Draft PR which builds a ARM64 container: https://github.com/developmentseed/titiler/pull/740
The PR needs work/refactoring, but here for comment.
